### PR TITLE
Possible 24 bytes additional savings

### PR DESCRIPTION
**Pull request by 640-KB**
Some of these may have potential impacts to the code that I may not be aware of, so I will document and comment them so we can strike any that might be problematic.

### CODE REVIEW

File: Lasse/LittleWindows.asm [hunk 1]
```diff
@@ -177,7 +177,8 @@ matched:
 ; we use the ror13 hash for the name of the API functions to not push the whole string of the api onto the stack,
 ; example: https://medium.com/asecuritysite-when-bob-met-alice/ror13-and-its-linkage-to-api-calls-within-modules-c2191b35161d
 resolve_symbols_kernel32:
-        mov edi, 10h                             ; edi will be used as an index to where on ebp the function address will be stored
```
Comment (640-KB) [inline Lasse/LittleWindows.asm:180]: -2 bytes shorter than `mov r32, imm`

File: Lasse/LittleWindows.asm [hunk 2]
```diff
@@ -202,7 +203,7 @@ load_user32:                                    ; Push "user32.dll" onto the sta
         call dword ptr[ebp + fn_LoadLibraryA]
 
 resolve_symbols_user32:
-        mov ebx, eax                            ; Load the base address of user32.dll into ebx
+        xchg eax, ebx                           ; Load the base address of user32.dll into ebx
```
Comment (640-KB) [inline Lasse/LittleWindows.asm:206]: -1 byte *if* it's okay to clobber eax here.

File: Lasse/LittleWindows.asm [hunk 3]
```diff
@@ -254,7 +255,7 @@ load_gdi32:                                     ; push "gdi32.dll" onto the stac
         call dword ptr[ebp + fn_LoadLibraryA]
 
 resolve_symbols_gdi32:
-        mov ebx, eax                            ; ebx = base address of gdi32.dll
+        xchg eax, ebx                           ; ebx = base address of gdi32.dll
```
Comment (640-KB) [inline Lasse/LittleWindows.asm:258]: -1 byte if it's okay to clobber eax here.

File: Lasse/LittleWindows.asm [hunk 4]
```diff
@@ -279,8 +280,7 @@ MainEntry:
         push esp                                ; Pointer to struct
         call dword ptr[ebp + fn_GetStartupInfoA]
         lea eax, (STARTUPINFOA ptr[esp]).wShowWindow
-        mov ecx, 1
-        test eax, ecx                           ; Find out if wShowWindow should be used
+        test al, 1                              ; Find out if wShowWindow should be used
```
Comment (640-KB) [inline Lasse/LittleWindows.asm:283]: I'm not clear if there's a need to set `ecx` to `1` right here, so if you're only using it to check the low bit of `eax` then a `test al, 1` should be all you need.  

If you do need `ecx` to be `1` hereafter, then `xor ecx, ecx`, `inc ecx` will still save 2 bytes.

File: Lasse/LittleWindows.asm [hunk 5]
```diff
@@ -422,9 +422,7 @@ WndProc:
```
Comment (640-KB) [inline Lasse/LittleWindows.asm:422]: consolidating leave stack frame and `ret 10h` saves a few more

**Review by 640-KB — COMMENTED** [review]
Comments on the changes are inline.

**Comment by rbergen** [comment]
Have you established that the executable produced with these changes starts, runs and behaves the way it should? That is basically the (only) criterion that has to be met for any version of the code.

**Comment by 640-KB** [comment]
> Have you established that the executable produced with these changes starts, runs and behaves the way it should? That is basically the (only) criterion that has to be met for any version of the code.

Yes, I tested this version (1,204 bytes using LINK) vs the previous (1,232 bytes) and I see no differences in behavior.  They both paint, resize, min, max, close, text centered, title bar, system menu... and most importantly - doesn't crash!

**Comment by 640-KB** [comment]
> Have you established that the executable produced with these changes starts, runs and behaves the way it should? That is basically the (only) criterion that has to be met for any version of the code.

Here's a question though...  when it comes to byte golfing, what is our "cost of victory"?

For example, would we be willing to trade:

```
    68 00 00 00 80     push CW_USEDEFAULT          ; y (0x80000000)
    68 00 00 00 80     push CW_USEDEFAULT          ; x
```

for:

```
    99                 cdq                         ; edx = 0
    F9                 stc                         ; cf = 1
    D1 DA              rcr edx, 1                  ; edx = CW_USEDEFAULT (0x80000000)
    52                 push edx                    ; y
    52                 push edx                    ; x
```

... to save 4 bytes?

**Comment by rbergen** [comment]
I'd say so. As Dave put it in the README, the request is to _"keep it readable **and explain what you're doing in the comments**"_. I think the key is in the bold part. For instance, like this:

```asm
    99                 cdq                         ; edx = 0
    F9                 stc                         ; set carry flag
    D1 DA              rcr edx, 1                  ; rotate carry flag into leftmost bit of edx
                                                   ; edx is now equal to CW_USEDEFAULT (0x80000000)
    52                 push edx                    ; y
    52                 push edx                    ; x 
```

Because also: _"the smaller, the better!"_

**Comment by rbergen** [comment]
Looking at the size of the executables (all of which work), this is what I see:

| Linker | main branch | this PR |
|-|-:|-:|
| LINK with `/ALIGN:16` | 1216 | 1232 |
| LINK with `/ALIGN:4` | 1220 | 1204 |
| Crinkler | 864 | 857 |

**Comment by 640-KB** [comment]
> I'd say so. As Dave put it in the README, the request is to _"keep it readable **and explain what you're doing in the comments**"_.  Because also: _"the smaller, the better!"_

Got it.  In that case I may also include a comment about why we're doing this vs. the `push imm` so that's clear as well.

**Comment by 640-KB** [comment]
`LINK with /ALIGN:4` with this updated PR now builds for me at 1,196 (1,194 code size).

**Comment by rbergen** [comment]
Does this conclude the set of changes you now propose to make?

**Comment by 640-KB** [comment]
> Does this conclude the set of changes you now propose to make?

Yes, done (for now).

**Comment by 640-KB** [comment]
Sorry, this time the last for real!

`LINK with /ALIGN:4` with this updated PR now builds for me at 1,188.

**Comment by 640-KB** [comment]
Okay, don't shoot me... but I've got a 1,108 EXE and I think I can do a good bit more still.  It's a separate PR for sure, so this one is still clear as far as I'm concerned.

Stay tuned!

**Comment by rbergen** [comment]
Don't worry. I've been managing the Primes repository for well over a year and a half now, which is another project with an optimization objective. I've seen my fair share of PRs with endless sequences of "no, really, this is the last one" commits. :)

Anyway, I can confirm the 1188 byte executable (or 854 with Crinkler), and the fact it works as well.

With that being the state of affairs, I'll merge this and update the "current sizes" section in the README.

**Pull request merged**